trin-core/src/portalnet/storage.rs

```diff
@@ -24,82 +26,217 @@ pub enum DistanceFunction {
+            Self::Io(err) => format!("IO error {}", err),
+            Self::InsufficientRadius { radius, distance } => format!(
+                "radius {} insufficient to store content at distance {}",
+                radius, distance
+            ),
+            Self::InvalidData(err) => format!("data invalid {}", err),
+        };
+
+        write!(f, "{}", message)
+    }
+}
+
+impl std::error::Error for PortalContentStoreError {}
+
+impl From<rocksdb::Error> for PortalContentStoreError {
+    fn from(err: rocksdb::Error) -> Self {
+        Self::Database(format!("(rocksdb) {}", err.to_string()))
+    }
+}
+
+impl From<rusqlite::Error> for PortalContentStoreError {
+    fn from(err: rusqlite::Error) -> Self {
+        Self::Database(format!("(sqlite) {}", err.to_string()))
+    }
+}
+
+impl From<r2d2::Error> for PortalContentStoreError {
+    fn from(err: r2d2::Error) -> Self {
+        Self::Database(format!("(r2d2) {}", err.to_string()))
+    }
+}
+
+impl From<std::io::Error> for PortalContentStoreError {
+    fn from(err: std::io::Error) -> Self {
+        Self::Io(err)
+    }
+}
+
+/// A data store for Portal Network content (data).
+pub trait PortalContentStore {
+    /// Looks up a piece of content by `key`.
+    fn get<K: OverlayContentKey>(
+        &self,
+        key: &K,
+    ) -> Result<Option<Vec<u8>>, PortalContentStoreError>;
+
+    /// Puts a piece of content into the store.
+    fn put<K: OverlayContentKey, V: AsRef<[u8]>>(
+        &mut self,
+        key: K,
+        value: V,
+    ) -> Result<(), PortalContentStoreError>;
+
+    /// Returns whether the content denoted by `key` is within the radius of the data store.
+    fn is_key_within_radius<K: OverlayContentKey>(&self, key: &K) -> bool;
+
+    /// Returns the radius of the data store.
+    fn radius(&self) -> U256;
+}
+
+/// An in-memory `PortalContentStore`.
+pub struct MemoryPortalContentStore {
+    /// The content store.
+    store: std::collections::HashMap<Vec<u8>, Vec<u8>>,
+    /// The `NodeId` of the local node.
+    node_id: NodeId,
+    /// The distance function used by the store to compute distances.
+    distance_fn: DistanceFunction,
+    /// The radius of the store.
+    radius: U256,
+}
+
+impl MemoryPortalContentStore {
+    /// Constructs a new `MemoryPortalContentStore`.
+    pub fn new(node_id: NodeId, distance_fn: DistanceFunction) -> Self {
+        Self {
+            store: std::collections::HashMap::new(),
+            node_id,
+            distance_fn,
+            radius: U256::MAX,
+        }
+    }
+
+    /// Returns the distance to `key` from the local `NodeId` according to the distance function.
+    fn distance_to_key<K: OverlayContentKey>(&self, key: &K) -> U256 {
+        match self.distance_fn {
+            DistanceFunction::Xor => XorMetric::distance(&key.content_id(), &self.node_id.raw()),
+        }
+    }
+}
+
+impl PortalContentStore for MemoryPortalContentStore {
```

Comment:
I'd prefer to see `MemoryPortalStorage` here. I thought the ideas was that it's a like-for-like replacement for `PortalStorage`?

Reply:
I'm not sure what you mean. Are you asking why the name is not `MemoryPortalStorage`?

Reply:
Yeah pretty much, but you can ignore this if we're going with the `"...Store"` naming scheme